okhttp/src/main/kotlin/okhttp3/OkHttpClient.kt

```diff
@@ -224,30 +222,46 @@ open class OkHttpClient internal constructor(
   constructor() : this(Builder())
 
   init {
-    if (builder.sslSocketFactoryOrNull != null || connectionSpecs.none { it.isTls }) {
+    if (connectionSpecs.none { it.isTls }) {
+      this.sslSocketFactoryOrNull = null
+      this.certificateChainCleaner = null
+      this.x509TrustManager = null
+      this.certificatePinner = CertificatePinner.DEFAULT
+    } else if (builder.sslSocketFactoryOrNull != null) {
       this.sslSocketFactoryOrNull = builder.sslSocketFactoryOrNull
-      this.certificateChainCleaner = builder.certificateChainCleaner
-      this.x509TrustManager = builder.x509TrustManagerOrNull
+      this.certificateChainCleaner = builder.certificateChainCleaner!!
+      this.x509TrustManager = builder.x509TrustManagerOrNull!!
+      this.certificatePinner = builder.certificatePinner
+          .withCertificateChainCleaner(certificateChainCleaner!!)
     } else {
       this.x509TrustManager = Platform.get().platformTrustManager()
-      Platform.get().configureTrustManager(x509TrustManager)
-      this.sslSocketFactoryOrNull = newSslSocketFactory(x509TrustManager!!)
+      this.sslSocketFactoryOrNull = Platform.get().newSslSocketFactory(x509TrustManager!!)
       this.certificateChainCleaner = CertificateChainCleaner.get(x509TrustManager!!)
+      this.certificatePinner = builder.certificatePinner
+          .withCertificateChainCleaner(certificateChainCleaner!!)
     }
 
-    if (sslSocketFactoryOrNull != null) {
-      Platform.get().configureSslSocketFactory(sslSocketFactoryOrNull)
-    }
-
-    this.certificatePinner = builder.certificatePinner
-        .withCertificateChainCleaner(certificateChainCleaner)
+    verifyClientState()
+  }
 
+  private fun verifyClientState() {
     check(null !in (interceptors as List<Interceptor?>)) {
       "Null interceptor: $interceptors"
     }
     check(null !in (networkInterceptors as List<Interceptor?>)) {
       "Null network interceptor: $networkInterceptors"
     }
+
+    if (connectionSpecs.none { it.isTls }) {
+      check(sslSocketFactoryOrNull == null)
```

Comment:
This assertion isn’t very interesting! Seems impossible to violate.

Reply:
Maybe omit these 4 checks, and invert the if ?
```
     if (connectionSpecs.any { it.isTls }) { 
```

Reply:
OK, will check other case.  I think I had this mainly for testing, I want to leave it here because carrying these through weird chains of clients, is I think how we got into the wacky state with NPE.